extension/src-language-server/snippets/stpa-snippets.ts

```diff
@@ -0,0 +1,382 @@
+ */
+function getPositionForCSSnippet(document: TextDocument, snippet: LanguageSnippet): Position {
+    const docText = document.getText();
+
+    // determine range of already existing definition of control structure
+    const titleIndex = docText.indexOf("ControlStructure");
+    const startIndex = titleIndex !== -1 ? titleIndex : 0;
+    const nextTitleIndex = docText.indexOf("Responsibilities");
+    const endIndex = nextTitleIndex !== -1 ? nextTitleIndex - 1 : docText.length - 1;
+    if (titleIndex === -1) {
+        return document.positionAt(endIndex);
+    } else {
+        // delete the control structure keyword
+        snippet.insertText = snippet.insertText.substring(18, snippet.insertText.length);
+        // check whether a graph ID already exist
+        const csText = docText.substring(startIndex, endIndex + 1);
+        const graphIndex = csText.indexOf("{");
+        if (graphIndex === -1) {
+            return document.positionAt(endIndex);
+        } else {
+            // delete the graph ID
+            snippet.insertText = snippet.insertText.substring(
+                snippet.insertText.indexOf("{") + 1,
+                snippet.insertText.lastIndexOf("}")
+            );
+            const bracketIndex = csText.lastIndexOf("}");
+            return document.positionAt(titleIndex + bracketIndex);
+        }
+    }
+}
+// TODO: maybe need adjustments since CS can be nested now
+/**
+ * Adds {@code id} to the control structure node names in {@code text}.
+ * @param text The control structure text.
+ * @param id The id to append.
+ * @returns The modified text.
+ */
+function addNodeIDs(text: string, id: string): string {
+    const splits = text.split(/[^a-zA-Z0-9\{\}]/);
+    // collect node names
+    const names: string[] = [];
+    for (let i = 3; i < splits.length; i++) {
+        if (splits[i] === "{" && !isKeyWord(splits[i - 1])) {
+            names.push(splits[i - 1]);
+        }
+    }
+
+    // append ID to node names
+    for (const name of names) {
+        const regex = new RegExp(name, "g");
+        text = text.replace(regex, name + id);
+    }
+    return text;
+}
+
+/**
+ * Checks whether the given {@code text} is a keyword.
+ * @param text The text to check.
+ * @returns whether the given {@code text} is a keyword.
+ */
+function isKeyWord(text: string): boolean {
+    return (
+        text === "hierarchyLevel" ||
+        text === "label" ||
+        text === "processModel" ||
+        text === "controlActions" ||
+        text === "feedback"
+    );
+}
+
+/**
+ * Snippet language definition for STPA.
+ */
+export class STPALanguageSnippet implements LanguageSnippet {
+    /** The text that should be inserted when clicking on the snippet */
+    insertText: string;
+    documents: LangiumDocuments;
+    /** The id of the snippet */
+    id: string;
+    /** Counts how many times the snippet was added already  */
+    protected counter: number = 0;
```

Comment:
why does this need to be counted?

Reply:
what if I manually write the code that the snipped would add? snippets shouldn't depend on the number of times they are clicked